hbase-endpoint/src/test/java/org/apache/hadoop/hbase/coprocessor/TestAsyncRegionServersCoprocessorEndpoint.java

```diff
@@ -0,0 +1,137 @@
+    HBaseClassTestRule.forClass(TestAsyncRegionServersCoprocessorEndpoint.class);
+
+  private static final FileNotFoundException WHAT_TO_THROW = new FileNotFoundException("/file.txt");
+  private static final String DUMMY_VALUE = "val";
+  private static final int NUM_SLAVES = 5;
+
+  @BeforeClass
+  public static void setUpBeforeClass() throws Exception {
+    TEST_UTIL.getConfiguration().setInt(HConstants.HBASE_RPC_TIMEOUT_KEY, 60000);
+    TEST_UTIL.getConfiguration().setInt(HConstants.HBASE_CLIENT_OPERATION_TIMEOUT, 120000);
+    TEST_UTIL.getConfiguration().setInt(HConstants.HBASE_CLIENT_RETRIES_NUMBER, 2);
+    TEST_UTIL.getConfiguration().setStrings(CoprocessorHost.MASTER_COPROCESSOR_CONF_KEY,
+      ProtobufCoprocessorService.class.getName());
+    TEST_UTIL.getConfiguration().setStrings(CoprocessorHost.REGIONSERVER_COPROCESSOR_CONF_KEY,
+      DummyRegionServerEndpoint.class.getName());
+    TEST_UTIL.startMiniCluster(NUM_SLAVES);
+    ASYNC_CONN = ConnectionFactory.createAsyncConnection(TEST_UTIL.getConfiguration()).get();
+  }
+
+  @Test
+  public void testRegionServersCoprocessorService()
+    throws ExecutionException, InterruptedException {
+    final List<JVMClusterUtil.RegionServerThread> regionServerThreads =
+      TEST_UTIL.getHBaseCluster().getRegionServerThreads();
+    List<ServerName> serverNames = new ArrayList<>();
+    regionServerThreads.forEach(t -> serverNames.add(t.getRegionServer().getServerName()));
+
+    DummyRequest request = DummyRequest.getDefaultInstance();
+    Map<ServerName, Object> resultMap =
+      admin.<DummyService.Stub, DummyResponse> coprocessorService(DummyService::newStub,
+        (s, c, done) -> s.dummyCall(c, request, done), serverNames).get();
+
+    resultMap.forEach((k, v) -> {
+      assertTrue(v instanceof DummyResponse);
+      DummyResponse resp = (DummyResponse) v;
+      assertEquals(DUMMY_VALUE, resp.getValue());
+    });
+  }
+
+  @Test
+  public void testRegionServerCoprocessorsServiceError()
+    throws ExecutionException, InterruptedException {
+    final List<JVMClusterUtil.RegionServerThread> regionServerThreads =
+      TEST_UTIL.getHBaseCluster().getRegionServerThreads();
+    List<ServerName> serverNames = new ArrayList<>();
+    regionServerThreads.forEach(t -> serverNames.add(t.getRegionServer().getServerName()));
+
+    DummyRequest request = DummyRequest.getDefaultInstance();
+    Map<ServerName, Object> resultMap =
+      admin.<DummyService.Stub, DummyResponse> coprocessorService(DummyService::newStub,
+        (s, c, done) -> s.dummyThrow(c, request, done), serverNames).get();
+
+    resultMap.forEach((k, v) -> {
+      assertTrue(v instanceof RetriesExhaustedException);
+      Throwable e = (Throwable) v;
+      assertTrue(e.getMessage().contains(WHAT_TO_THROW.getClass().getName().trim()));
+    });
+  }
+
+  public static class DummyRegionServerEndpoint extends DummyService
+    implements RegionServerCoprocessor {
```

Comment:
will check to see here if we can make it work like: the execution is successful on 3/5 servers and fails on 2/5 servers. maybe we can introduce atomic int that we keep incrementing with each dummyCall() and when count > 3, we return error (something like that)